File: legacy/ts-cli/src/core/readiness/scoring.ts
import type { CheckResult, CheckSeverity, ReadinessLevel } from './types.ts'

const PENALTIES: Record<CheckSeverity, number> = {
	critical: 20,
	high: 10,
	medium: 5,
	low: 2,
	info: 0,
}

export function scoreChecks(checks: CheckResult[]): number {
	const penalty = checks.reduce((total, check) => {
		if (check.status !== 'failed') {
			return total
		}

		return total + PENALTIES[check.severity]
	}, 0)

	return Math.max(0, Math.min(100, 100 - penalty))
}

export function getReadinessLevel(score: number): ReadinessLevel {
	if (score >= 90) return 'production-ready'
	if (score >= 75) return 'almost-ready'
	if (score >= 50) return 'risky'
	return 'not-production-ready'
}

export function formatReadinessLevel(level: ReadinessLevel): string {
	switch (level) {
		case 'production-ready':
			return 'Production Ready'
		case 'almost-ready':
			return 'Almost Ready'
		case 'risky':
			return 'Risky'
		case 'not-production-ready':
			return 'Not Production Ready'
	}
}

File: go-cli/internal/readiness/scoring.go
package readiness

var penalties = map[Severity]int{
	Critical: 20,
	High:     10,
	Medium:   5,
	Low:      2,
	Info:     0,
}

func Score(checks []CheckResult) int {
	total := 100
	for _, check := range checks {
		if check.Status == Failed {
			total -= penalties[check.Severity]
		}
	}
	if total < 0 {
		return 0
	}
	if total > 100 {
		return 100
	}
	return total
}

func ReadinessLevel(score int) Level {
	switch {
	case score >= 90:
		return "production-ready"
	case score >= 75:
		return "almost-ready"
	case score >= 50:
		return "risky"
	default:
		return "not-production-ready"
	}
}

func LevelLabel(level Level) string {
	switch level {
	case "production-ready":
		return "Production Ready"
	case "almost-ready":
		return "Almost Ready"
	case "risky":
		return "Risky"
	default:
		return "Not Production Ready"
	}
}
